Only quote search engine terms if requested

t-wissmann opened the pull request:
In the search engine url replacement, only quote the search term if the
search engine format explicitly requests this:

  - The placeholder {} inserts the plain search term without quotation
    like before of #4434. This seems to be the default for other
    browsers' search engines and to be supported by most websites. For
    example, https://en.wikipedia.org/wiki/AC/DC leads to the wikipedia
    article of AC/DC.
  - A new placeholder {quoted} inserts the quoted search term. For
    example, with

        c.url.searchengines['jpc'] = 'https://www.jpc.de/s/{quoted}'

    the command "jpc AC/DC" correctly redirects to https://www.jpc.de/s/AC%2FDC
    given results for "AC/DC". Here, the placeholder {quoted} is indeed
    necessary (as of March 2020), since  https://www.jpc.de/s/AC/DC only
    searches for "AC".

This addresses the issue #1772 and partially reverts #4434 (making the
there introduced behaviour configurable).

jb55 commented:
I like the approach :+1:  thank you for doing this

The-Compiler commented on qutebrowser/config/configtypes.py:
```diff
@@ -1760,11 +1760,14 @@ def to_py(self, value: _StrUnset) -> _StrUnsetNone:
+            format_keys = {
+                'quoted': "",
+            }
+            value.format("", **format_keys)
```
There already is a `FormatString` type - I wonder, maybe `SearchEngineUrl` should have a `FormatString` member (or inherit from it, though I prefer composition) and use that? That already has a way to check which placeholders are allowed, so you'd only need to extend it to have a `requires_placeholder=True` (or so) flag (it probably could just search for `{.*}` or something then).

t-wissmann replied:
For me (not knowing python and your code base well enough), this is too much work, because `FormatString` does not allow `{}`, so one also needs to take care of that.. (And I think this is something one can also do later)

The-Compiler commented on qutebrowser/config/configtypes.py:
```diff
@@ -1760,11 +1760,14 @@ def to_py(self, value: _StrUnset) -> _StrUnsetNone:
+            value.format("", **format_keys)
         except (KeyError, IndexError):
             raise configexc.ValidationError(
                 value, "may not contain {...} (use {{ and }} for literal {/})")
```
I can't comment on the proper line - but I'm pretty sure the `QUrl` check below would fail with a value only using `{quoted}` but not `{}`, no?

t-wissmann replied:
I'm not sure whether `QUrl` check still makes sense if we allow `{}`, `{quoted}` and `{unquoted}`.

t-wissmann commented:
> I'd prefer a different approach, for ~two~ three reasons:
> 
> * This change breaks everyone's existing configs - #4434 is in since November
>   2018, so I'd rather have a change which preserves the existing behavior for
>   `{}`
> * Maybe remove `safe=''` (i.e. revert #4434) for the default case, which would make the behavior in-line with Chromium (i.e. help with [#4434 (comment)](https://github.com/qutebrowser/qutebrowser/pull/4434#issuecomment-604679322) but not with #1772) - and be slightly backwards compatible.

I'm sorry, I've messed this up... You're absolutely right, `{}` must quote
almost all characters. I've updated this PR and removed the `safe=''`. As for
#4434 most users should not notice the difference, and for a few people, it
makes things work again and `{}` behaves like Chrome's placeholder.

There are now three placeholders (I've updated the documentation and the
testcases). By the way, should I run `scripts/dev/src2asciidoc.py` before
commiting?

The-Compiler commented:
> For me (not knowing python and your code base well enough), this is too much work, because `FormatString` does not allow `{}`, so one also needs to take care of that.. (And I think this is something one can also do later)

Fine by me. Probably not worth the trouble either way.

> I'm not sure whether `QUrl` check still makes sense if we allow `{}`, `{quoted}` and `{unquoted}`.

Hmm, right... I suppose we can remove it then?

> By the way, should I run `scripts/dev/src2asciidoc.py` before
> commiting?

Doesn't matter much to me - I usually do so while merging.

Looks like some [tests are failing](https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/670877873#L560) and [pylint has something to say](https://travis-ci.org/github/qutebrowser/qutebrowser/jobs/670877876) as well - could you take a look at those please?

t-wissmann commented:
I've removed the `QUrl` check and I'm still working on the remaining failed test cases. Sorry for the delay!

The-Compiler commented:
No worries - feel free to ping me if you need help!

t-wissmann commented:
Alright! The testcases are fixed now and the CI seems to be happy: I removed the testcase where the `QUrl`-syntax check was triggered and added a testcase `test_get_search_url_for_path_search()` for search engines that insert the search term in the path (instead of in the GET-parameter, i.e. the query).

So from my point of view, this is PR is ready. Take your time!

t-wissmann commented:
Thanks!

The-Compiler commented:
And thank you! I noticed there was an unneeded (at least IMHO) dictionary and also added a changelog entry:

- 408dc809c Remove unneeded dictionary
- 9540f5377 Update changelog
